### DISCUSSION TITLE
Save metrics, logs, traces locally?

### DISCUSSION THREAD

**Discussion by LoikaAR** [open-telemetry/opentelemetry-java#6342] [Q&A]
Might be a dumb question but is it absolutely required to set up an HTTP/gRPC server to recieve metrics, logs and traces? Is there a way to save them locally? I am quite new so tell me if I am thinking in the wrong direction.

**Comment by jack-berg** [comment]
The [opentelemetry-exporter-logging-otlp](https://github.com/open-telemetry/opentelemetry-java?tab=readme-ov-file#sdk-exporters) exporters will log the metrics, traces, and log records in a standard JSON encoding of the OTLP format. From here, you can configure your logging system to append them to files or however else you want to store them locally.

You can also write custom `MetricExporter`, `SpanExporter`, `LogRecordExporter` implementations which do exactly what you want. 

Or you can export to a local collector via OTLP, and use one of the [many exporters](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter) bundled with the collector like the [fileexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/fileexporter). Generally, there are more exporters implemented in the collector than directly in language SDKs since writing an exporter for the collector allows you to build it once instead of building it 10+ times for all the different OpenTelemetry language implementations.

**Reply by LoikaAR** [reply]
Thank you. I seem to be receiving metrics, but no logs or traces. Is this a common issue or should I open a new discussion?

**Reply by LoikaAR** [reply]
At this point I only need to collect the local variable info - their names/values - at runtime. If you think I am overdoing please tell me

**Reply by jack-berg** [reply]
The telemetry produced is a function of your application / dependencies, instrumentation setup, and SDK configuration. I.e. if you only have instrumentation or SDK configured to produce metrics, you wouldn't expect to see traces or logs. Most people use the otel java agent, which produces traces, metrics, and logs by default for a variety of popular libraries.

How are you using opentelemetry java? Are you manually instrumenting, or using library instrumentation from [opentelemetry-java-instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation), or using the otel java agent for auto instrument from [opentelemetry-java-instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation).

**Reply by LoikaAR** [reply]
Thanks for the reply. I am using the otel java agent like this:

`java -javaagent:/path/to/opentelemetry-javaagent.jar -jar /path/to/project.jar`

And in the end I've set up a very simple post request handler on localhost:4318, which just prints the response from receiving `/v1/metrics`, `/v1/traces` and `/v1/logs`. So far it seems to only receive metrics.

**Reply by jack-berg** [reply]
The otel java agent has OTLP export enabled by default for traces, metrics, and logs by default. If you're only seeing metrics, it could be that you're not using any of the supported libraries for traces or logs: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md

Or it could be that the app isn't being exercised in a way that produces trace or log telemetry.

This example shows a simple spring boot with the otel java agent emitting metric, log and tracing telemetry to an opentelemetry collector configured to log the data to stdout: https://github.com/open-telemetry/opentelemetry-java-examples/tree/main/javaagent

**Reply by LoikaAR** [reply]
I verified using your suggestions (thank you), but still no luck.

I am using plain Java 17 (with JaCoCo and Junit for coverage), with the following server code:

`public class OtelServerHandler {

    public static void main(String[] args) {
        port(4318);

        post("/v1/traces", (request, response) -> {
            String body = request.body();
            System.out.println("Received traces: " + body);
            return "Traces received";
        });

        post("/v1/metrics", (request, response) -> {
            String body = request.body();
            System.out.println("Received metrics: " + body);
            return "Metrics received";
        });

        post("/v1/logs", (request, response) -> {
            String body = request.body();
            System.out.println("Received logs: " + body);
            return "Logs received";
        });
    }
}
`

And my instrumented program is just a very simple algorithm that I am trying to get  logs/traces from.

**Reply by jack-berg** [reply]
Maybe try emitting some spans manually to verify? Can obtain an SDK `TracerProvider` installed by the agent with `GlobalOpenTelemetry.get().getTracerProvider()`, and follow the instructions here to produce spans: https://opentelemetry.io/docs/languages/java/instrumentation/#traces

If you still don't see any spans, consider posting a small reproduction to github and I can take a closer look. If you do see spans, then the issue is that the otel java agent isn't installing any instrumentation automatically based on the dependencies of your application.

**Reply by LoikaAR** [reply]
Thank you very much for your offer. I tried to emit spans manually and had some questionable success. As in now I can print the span info but it appears to be empty, so it is possible I am doing it wrong. I have it in this repo here:

`git@github.com:LoikaAR/APR_Patch_Reports.git` (or tell me if I should post code elsewhere)

Everything is in the APRS_Practise folder. If you are able to have a look I would be very grateful, meanwhile I will try to remake it using springboot.

**Reply by LoikaAR** [reply]
I also tried to run the example application you provided. When I exercise the application by

`curl http://localhost:8080/ping`

I get this error message, similar to my own mini project (just with logs instead of metrics) before I cleaned up my dependency jars:

`javaagent-app-1 | [otel.javaagent 2024-04-06 08:46:27:199 +0000] [OkHttp http://collector:4318/...] ERROR io.opentelemetry.exporter.internal.http.HttpExporter - Failed to export logs. The request could not be executed. Full error message: timeout`

May it be something wrong with my IntelliJ or my machine?
